### ISSUE TITLE
Default values not changing when dependencies updated

### ISSUE THREAD

**Issue by chopfitzroy** [rjsf-team/react-jsonschema-form#4325]
### Prerequisites

- [X] I have searched the existing issues
- [X] I understand that providing a [SSCCE](http://sscce.org/) example is tremendously useful to the maintainers.
- [X] I have read the [documentation](https://rjsf-team.github.io/react-jsonschema-form/docs)
- [X] Ideally, I'm providing a [sample JSFiddle](https://jsfiddle.net), [Codesandbox.io](https://codesandbox.io) or preferably a [shared playground link](https://rjsf-team.github.io/react-jsonschema-form/) demonstrating the issue.

### What theme are you using?

core

### Version

5.x

### Current Behavior

Trying to use `default` values with `dependencies` but the `formData` does not update.

Visually the form does update which is quite confusing for the user. Fortunately if the user tries to submit the validation does catch the error.

Interestingly if I use `if` / `then` the issue presents the same but I do see warnings in the console, example here: https://codesandbox.io/p/sandbox/nagaozen-react-jsonschema-form-playground-forked-8g34z7 **(this is different from the Codesandbox link below, please check both)**

A number of issues suggest that this may be to do with uncontrolled vs. controlled forms but I have tried with both.

- https://github.com/rjsf-team/react-jsonschema-form/issues/975
- https://github.com/rjsf-team/react-jsonschema-form/issues/1316
- https://github.com/rjsf-team/react-jsonschema-form/issues/1394
- https://github.com/rjsf-team/react-jsonschema-form/issues/1399

These issues may also be tangentially related but it is hard to tell:

- https://github.com/rjsf-team/react-jsonschema-form/issues/4322
- https://github.com/rjsf-team/react-jsonschema-form/issues/4208
- https://github.com/rjsf-team/react-jsonschema-form/issues/3892

### Expected Behavior

When changing from _Cat_ to _Fish_ I expect the `formData` to update `food` from `"meat"` to `"worms"`.

### Steps To Reproduce

Example here: https://codesandbox.io/p/sandbox/nagaozen-react-jsonschema-form-playground-forked-t8f767
- Select _Cat_
- Check console should see `{food: 'meat', animal: 'Cat'}`
- Select _Fish_
- Check console should see `{food: 'meat', water: 'sea', animal: 'Fish'}`
- Try to submit the form - validation should prevent this

NOTE: The inverse is also true if you start with _Fish_ and then select _Cat_ `food` will be `"worms"`

### Environment

```markdown
- OS: WSL 2 - OpenSuse Tumbleweed
- Node: v22.6.0
- npm: 10.8.2
```

**Comment by nickgros** [comment]
I expected this to work if we set the experimental_defaultFormStateBehavior to `{ allOf: 'populateDefaults' }` but that didn't work. Definitely a bug here. I think we have a few similar reports related to conditional logic, allOf, and computing default values.

**Comment by chopfitzroy** [comment]
@nickgros sorry about that! Link is public now.

**Comment by wslbxx123** [comment]
Hi @heat!h-freenome, we are facing the same issue now and hope this can be fixed soon. when can we expect the fix? many thx!

**Comment by trentfridey** [comment]
Any updates on this? PR #4388 for the fix looks ready to merge -- but if it's not, I would be willing to help solve this since it is blocking my team from using RJSF currently.

**Comment by wslbxx123** [comment]
很高兴收到你的来信～祝你每天开心！
